File: 387.py
import random
def power(x,y,p):
    res=1
    x%=p
    while y:
        if y&1:
            res=(res*x)%p
        y>>=1
        x=(x*x)%p
    return res
def millerTest(d,n):
    a = 2+random.randint(1,n)%(n-4)
    x = power(a,d,n)
    if x==1 or x==n-1:
        return True
    while d!=n-1:
        x=(x*x)%n
        d*=2
        if x==1:
            return False
        if x==n-1:
            return True
    return False
def isPrime(n,k):
    if n<=1 or n==4:
        return False
    if n<=3:
        return True
    d=n-1
    while d%2==0:
        d//=2
    for i in range(0,k):
        if millerTest(d,n) == False:
            return False
    return True
def f():
    ans=[0]
    MAX=10**14
    def find(n,digsum,strong):
        m=n*10
        s=digsum
        for i in range(10):
            if m>=MAX:
                break
            if strong and isPrime(m,4):
                ans[0]+=m
            if m%s == 0:
                find(m,s,isPrime(m//s,4))
            m+=1
            s+=1
    for i in range(1,10):
        find(i,i,False)
    return str(ans[0])
if __name__ =="__main__":
    print(f())

File: 387.cpp
#include<bits/stdc++.h>
#define ull unsigned long long
using namespace std;
ull power(ull x, ull y, ull p) 
{ 
    ull res = 1;
    x = x % p;
    while (y) 
    {
        if (y&1) 
            res = (res*x) % p; 
        y>>=1;
        x=(x*x) % p; 
    }
    return res;
} 
bool miillerTest(ull d, ull n) 
{ 
    ull a = 2 + rand() % (n - 4); 
    ull x = power(a, d, n); 
    if (x == 1  || x == n-1) 
       return true; 
    while (d != n-1) 
    { 
        x = (x * x) % n; 
        d *= 2; 
        if (x == 1)      return false; 
        if (x == n-1)    return true; 
    } 
    return false; 
} 
bool isPrime(ull n, ull k) 
{ 
    if (n <= 1 || n == 4)  return false; 
    if (n <= 3) return true; 
    ull d = n - 1;
    while (d % 2 == 0) 
        d /= 2;
    for (ull i = 0; i < k; i++) 
         if (!miillerTest(d, n)) 
              return false; 
    return true; 
}
int main()
{
    ull ans=0;
    map<ull,ull> strongv[20];
    vector<pair<ull,ull>> v;
    for(int i=1;i<=9;i++)
        v.push_back({i,i});
    int l=1;
    while(l++)
    {
        if(l==4)break;
        int n = v.size();
        for(int i=0;i<n;i++)
        {
            for(int j=0;j<=9;j++)
            {
                string s = to_string(v[i].first)+to_string(j);
                ull t=stoll(s);
                ull digsum=v[i].second+j;
                if(t%digsum==0)
                {
                    v.push_back({t,digsum});
                    if(isPrime(t/digsum,4))
                        strongv[l+1][t]++;
                }
                if(isPrime(t,4)&&strongv[l][v[i].first])
                    ans+=t;
            }
        }
        v.erase(v.begin(),v.begin()+n);
    }
    cout<<ans;
}
